Input date calendar

Wilker0725 opened the pull request:
**Changes**
Description of changes...

- delete branch

<!-- Remove if not required -->
**Changelog entry**

- [BREAKING | WARNING] Description of change in `DatePicker Calendar`

**Additional information**

- You may refer to this [ticket](https://www.figma.com/file/us90jOpWBahsqK2VAluPvD/BookingSG-Design-System?node-id=3360%3A23096&t=KL4vwV8Fmg0hKXlF-0)
- Any other information you would like to include

keithtxw commented on src/date-input/date-input.tsx:
```diff
@@ -24,21 +27,32 @@ export const DateInput = ({
     const [monthValue, _setMonthValue] = useState<string>("");
     const [yearValue, _setYearValue] = useState<string>("");
     const [currentFocus, _setCurrentFocus] = useState<FieldType>("none");
+    const [calendarManualInput, setCalendarManualInput] = useState<string>("");
```
I have been looking through and find that using a string requires splitting and checking, which is laborious. Also curious as to the issue of assigning to the existing `dayValue` etc states?

Is it that any selection of within the calendar is not confirmed until you press ok? If so, keeping it as an object/separate variables might be easier

keithtxw commented on src/date-input/date-input.tsx:
```diff
@@ -369,6 +680,18 @@ export const DateInput = ({
                     maxLength={2}
                     value={monthValue}
                     ref={monthInputRef}
+                    {...(hoverValue !== undefined && {
```
i find this hard to understand. Could you consider a helper function to do this instead? Then assign an undefined to the prop rather than conditionally adding.

keithtxw commented on src/date-picker/types.ts:
```diff
@@ -0,0 +1,86 @@
+
+export type VariantDay =
+    | "default"
+    | "nextMonth"
```
convention is kebab-case, so it'll be `next-month`

keithtxw commented on src/date-picker/types.ts:
```diff
@@ -0,0 +1,86 @@
+    // Input-specific Attributes
+    error?: boolean;
+    readOnly?: boolean;
+    disabledDates?: string[]; // [YYYY-MM-DD]
```
add inline documentation instead and some brief explanation would be helpful
```
/** specifies the list of disabled dates in the YYYY-MM-DD format */
```

keithtxw commented on src/date-picker/types.ts:
```diff
@@ -0,0 +1,86 @@
+export interface GenerateDayStatusValue {
+    variant: VariantDay;
+    dateStartWithYear: string;
+    isDisabled: boolean;
```
`disabled`

keithtxw commented on src/date-picker/types.ts:
```diff
@@ -0,0 +1,86 @@
+
+export type View = "Day" | "Month" | "Year";
+
+export type FocusToTypes = {
```
I think we should use `interface` for consistency

keithtxw commented on src/date-picker/date-picker.tsx:
```diff
@@ -0,0 +1,1311 @@
+import React, {
```
This component is too big already. Can you consider pulling the calendar out instead?

keithtxw commented on src/date-picker/date-picker.tsx:
```diff
@@ -0,0 +1,1311 @@
+            range.length &&
+            dayjs(date).isBetween(start, range, "day", "[]")
+        ) {
+            classes.push("selected-between pre-selected-between");
```
I'm not comfortable about interchanging usage of class names for styled components. Would usage of props not suffice?

keithtxw reviewed (changes requested):
Perhaps we should spend some time to resolve fundamental concerns such as

- possibly extracting the calendar component and make it extensible to the desktop variant too
- type naming

Wilker0725 commented:
Work map flow:

- [ ] Clean up the date-input.tsx (Copy from master)
- [ ] Create move main input fn to new Component StandAloneInput
- [ ] Remove all date-picker files
- [ ] Remove date-range-input files
- [ ] onChange return value as string in date-input-tsx
- [ ] StandAloneInput sharable in range input (name="start-day" || name="end-day" )
- [ ] StandAloneInput's fn sharable in range input
- [ ] Rebase from master and Import Calendar
- [ ] Styling the calendar 
- [ ] Handle the interaction for start and end `StandAloneInput` in the calendar
- [ ] Styling the start and end Calendar
- [ ] Handle the interaction for start and end `StandAloneInput` from input itself

Note to this PR
1. date-input is the one handle final onChange to user.
2. date-input receive value from `StandAloneInput`  and `Calendar` Component

Wilker0725 commented:
Updated:
1. Uppercase the `INVALID_VALUE= "Invalid Date"` from `Invalid date`.
2. Add `performOnBlurHandler` and `performOnChangeHandler` inside `date-input-tsx`. return value {start: "DD-MM-YYYY"} or `{ start: "Invalid Date"}`.
3. Updated those `onChange`/`onChangeRaw`/`onBlur`/`onBlurRaw` => `ChangeValueTypes`

Wilker0725 commented:
Row alignment been applied in `375px` mobile devices which suggested from Selig. He also aware our Design System not exist `374px` but that is the best user experience in those `375px` devices.
